### PR TITLE
Make ~RSocketConnectionManager asynchronous, rename to ConnectionSet

### PR DESCRIPTION
**Pull request by alexmalyshev**
RSocketStateMachine now has a weak pointer to the ConnectionSet that holds it.
When the state machine is closed, it'll attempt to remove itself from the set.
When the set is about to be destroyed, it schedules RSocketStateMachine::close()
on all of its state machines, and exits immediately.

### CODE REVIEW

File: rsocket/RSocketServer.cpp [hunk 1]
```diff
@@ -143,7 +142,10 @@ void RSocketServer::onRSocketSetup(
       std::move(connectionParams.connectionEvents),
       nullptr, /* resumeManager */
       nullptr /* coldResumeHandler */);
-  connectionManager_->manageConnection(rs, *eventBase);
+
+  connectionSet_->insert(rs, eventBase);
+  rs->registerSet(connectionSet_);
```
Comment (alexmalyshev) [inline rsocket/RSocketServer.cpp]: I could fold all these calls to `registerSet()` into `ConnectionSet::insert()`, but that'd mean `ConnectionSet` would have to inherit from `enable_shared_from_this`, something which I don't really feel like doing....

File: rsocket/RSocketClient.h [hunk 2]
```diff
@@ -72,7 +72,7 @@ class RSocketClient {
   void createState(folly::EventBase& eventBase);
 
   std::shared_ptr<ConnectionFactory> connectionFactory_;
-  std::unique_ptr<RSocketConnectionManager> connectionManager_;
+  std::shared_ptr<ConnectionSet> connectionSet_;
```
Comment (phoad) [inline rsocket/RSocketClient.h]: The forward declaration of RSocketConnectionManager is still there.

**Comment by phoad** [comment]
This has resolved a bug in our codebase already: T21893729
💯 👍

File: rsocket/internal/ConnectionSet.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+ConnectionSet::ConnectionSet() {}
+
+ConnectionSet::~ConnectionSet() {
+  VLOG(1) << "Started ~ConnectionSet";
+  SCOPE_EXIT { VLOG(1) << "Finished ~ConnectionSet"; };
+
+  StateMachineMap map;
+
+  // Set up the baton and counter of connections yet to be closed.  Move all the
```
Comment (somasun) [inline rsocket/internal/ConnectionSet.cpp]: Stale comments ?

File: rsocket/internal/ConnectionSet.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  {
+    auto locked = machines_.lock();
+    if (locked->empty()) {
+      VLOG(2) << "No connections to close, early exit";
+      return;
+    }
+
+    VLOG(2) << "Need to close " << locked->size() << " connections";
+
+    map.swap(*locked);
```
Comment (somasun) [inline rsocket/internal/ConnectionSet.cpp]: Now that we are not waiting for all the stateMachines to close, can we get rid of the copy.  The code can become cleaner (and the destruction will also be faster)

Comment (alexmalyshev) [reply rsocket/internal/ConnectionSet.cpp]: Technically this is a move and not a copy.  This way should be faster, as other threads running `RSocketStateMachine::close()` can interleave with `~ConnectionSet`, as the `~ConnectionSet` thread isn't holding the lock.

**Pull request merged**